File: internal/port/vectorstore.go
package port

import "context"

// VectorQuery is a KNN search against the vector store. Filters are
// applied via JOIN on context_item — see sqlite impl.
type VectorQuery struct {
	Vector []float32
	Model  string // slug, must match embedding_model.slug
	Limit  int
	// Filters pushed down to context_item (same semantics as ItemFilter
	// for the same fields). Empty = no filter.
	Scopes []string
	Kinds  []string
}

// VectorHit is one KNN result.
type VectorHit struct {
	ID       string
	Score    float64 // higher = better (caller-normalized from distance)
	Distance float64 // raw vec0 distance (lower = better)
}

// VectorStore reads and writes embeddings keyed by item_id. A given
// item_id has at most one embedding per model (PRIMARY KEY in
// context_embedding).
type VectorStore interface {
	// Put writes (or replaces) the embedding for item_id under the
	// given model. Idempotent.
	Put(ctx context.Context, model, itemID string, vector []float32) error
	// Search runs a KNN query. Returns hits sorted by Score DESC.
	Search(ctx context.Context, q VectorQuery) ([]VectorHit, error)
	// Delete removes the embedding for item_id under model. No-op if absent.
	Delete(ctx context.Context, model, itemID string) error
}

File: python/src/unictx/search/vectorstore.py
"""VectorStore Protocol + VectorQuery/VectorHit.

Ports Go's internal/port/vectorstore.go. The vec0 virtual table is
owned by this Protocol's impl (Phase 2, storage/) — vector writes live
here, not in ContextRepo or EmbeddingRepo.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol, runtime_checkable


@dataclass(slots=True)
class VectorQuery:
    """KNN search against the vector store. Mirrors Go's port.VectorQuery.

    Filters (scopes, kinds) are pushed down to context_item via JOIN in
    the sqlite impl. Empty list = no filter on that dimension.

    Note on `vector` type: Go uses `[]float32`. We use `list[float]` —
    concrete and matches Go's concreteness. `Sequence[float]` was
    considered but rejected because it would accept tuples/iterators
    that the impl likely needs to index/slice, and we want callers to
    pass a real list.
    """

    vector: list[float] = field(default_factory=list)
    model: str = ""  # slug, must match embedding_model.slug
    limit: int = 0
    scopes: list[str] = field(default_factory=list)
    kinds: list[str] = field(default_factory=list)


@dataclass(slots=True)
class VectorHit:
    """One KNN result. Mirrors Go's port.VectorHit.

    score is caller-normalized from distance — higher = better.
    distance is the raw vec0 distance — lower = better. Both are
    returned because different callers want different signals.
    """

    id: str = ""
    score: float = 0.0
    distance: float = 0.0


@runtime_checkable
class VectorStore(Protocol):
    """Reads and writes embeddings keyed by item_id. Mirrors Go's port.VectorStore.

    A given (model, item_id) has at most one embedding — PRIMARY KEY in
    context_embedding. `put` is idempotent (insert-or-replace). `delete`
    is no-op if absent.
    """

    def put(self, model: str, item_id: str, vector: list[float]) -> None: ...

    def search(self, q: VectorQuery) -> list[VectorHit]:
        """KNN query. Returns hits sorted by score DESC."""
        ...

    def delete(self, model: str, item_id: str) -> None: ...
